deps/thpool/thpool.c

```diff
@@ -346,25 +456,186 @@ void redisearch_thpool_destroy(redisearch_thpool_t* thpool_p) {
     thread_destroy(thpool_p->threads[n]);
   }
   rm_free(thpool_p->threads);
+  rm_free(thpool_p->name);
   rm_free(thpool_p);
 }
 
 /* Pause all threads in threadpool */
 void redisearch_thpool_pause(redisearch_thpool_t* thpool_p) {
+  if (!thpool_p) return;
+
+  while (thpool_p->flags & RS_THPOOL_F_RESUME) {
+    sleep(1);
+    RedisModule_Log(
+        NULL, "warning",
+        "redisearch_thpool_pause(): waiting for %s threadpool to finish resuming process.",
+        thpool_p->name);
+  }
   size_t n;
-  for (n = 0; n < thpool_p->num_threads_alive; n++) {
-    pthread_kill(thpool_p->threads[n]->pthread, SIGUSR2);
+	pthread_t caller = pthread_self();
+
+  // save the current thpool to the global pointer
+  g_curr_threadpool = thpool_p;
+
+  size_t threadpool_size = thpool_p->num_threads_alive;
+
+  int called_by_threadpool = 0;
+
+  // set hold flag
+  thpool_p->flags |= RS_THPOOL_F_PAUSE;
+
+  // zero threads' ids. This number will be increased atomically by each
+  // thread so it can write to its own unique idx in the output array.
+  g_threads_ids = 0;
+
+	for(n = 0; n < threadpool_size; n++) {
+		// do not pause caller
+		if(thpool_p->threads[n]->pthread != caller) {
+			pthread_kill(thpool_p->threads[n]->pthread, SIGUSR2);
```

Comment:
This is signaling the thread to call `thread_hold`, right?

Reply:
Correct